#Allways enforce symlink checking
--- a/server/request.c
+++ b/server/request.c
@@ -485,7 +485,7 @@
     savename = (lfi->valid & APR_FINFO_NAME) ? lfi->name : NULL;
 
     /* if OPT_SYM_OWNER is unset, we only need to check target accessible */
-    if (!(opts & OPT_SYM_OWNER)) {
+    if (/*!(opts & OPT_SYM_OWNER)*/0) { // Always enforce symlink checks
         if (apr_stat(&fi, d, lfi->valid & ~(APR_FINFO_NAME | APR_FINFO_LINK), p)
             != APR_SUCCESS)
         {
@@ -1126,6 +1126,7 @@
              * if...we have allowed symlinks
              * skip the lstat and dummy up an APR_DIR value for thisinfo.
              */
+            /* Skip this... We always want symlinks checked
             if (r->finfo.filetype != APR_NOFILE
 #ifdef CASE_BLIND_FILESYSTEM
                 && (filename_len <= canonical_len)
@@ -1137,7 +1138,7 @@
                 ++seg;
                 continue;
             }
-
+            */
             /* We choose apr_stat with flag APR_FINFO_LINK here, rather that
              * plain apr_stat, so that we capture this path object rather than
              * its target.  We will replace the info with our target's info
