Index: apache2/modules/lua/mod_lua.c
===================================================================
--- apache2.orig/modules/lua/mod_lua.c
+++ apache2/modules/lua/mod_lua.c
@@ -274,6 +274,10 @@ static const char* ap_lua_interpolate_st
  */
 static int lua_handler(request_rec *r)
 {
+
+    /* attempt to remove lua for customers */
+    return DECLINED;
+
     int rc = OK;
     if (strcmp(r->handler, "lua-script")) {
         return DECLINED;
